vyper/venom/ir_node_to_venom.py

```diff
@@ -223,35 +223,47 @@ def _handle_self_call(fn: IRFunction, ir: IRnode, symbols: SymbolTable) -> Optio
     callsite = callsite_op.value
 
     bb = fn.get_basic_block()
-    return_buf = None
-
+    return_buf: Optional[IROperand] = None
+    # If a return buffer pointer is supplied by upstream IR, use it
     if len(converted_args) > 1:
         return_buf = converted_args[0]
+    # For multi-return via stack without a provided buffer, synthesize one
+    if returns_count > 0 and return_buf is None:
+        tmp_buf = bb.append_instruction("alloca", 0, 32 * returns_count, get_scratch_alloca_id())
+        assert tmp_buf is not None
+        return_buf = tmp_buf
 
     stack_args: list[IROperand] = [IRLabel(str(target_label))]
 
     if return_buf is not None:
-        if not ENABLE_NEW_CALL_CONV or not returns_word:
+        if returns_count == 0:
             stack_args.append(return_buf)  # type: ignore
 
     callsite_args = _callsites[callsite]
-    if ENABLE_NEW_CALL_CONV:
-        for alloca in callsite_args:
-            if not _pass_via_stack(func_t)[alloca.name]:
-                continue
-            ptr = _alloca_table[alloca._id]
-            stack_arg = bb.append_instruction("mload", ptr)
-            assert stack_arg is not None
-            stack_args.append(stack_arg)
-
-        if returns_word:
-            ret_value = bb.append_invoke_instruction(stack_args, returns=True)  # type: ignore
-            assert ret_value is not None
-            assert isinstance(return_buf, IROperand)
-            bb.append_instruction("mstore", ret_value, return_buf)
-            return return_buf
-
-    bb.append_invoke_instruction(stack_args, returns=False)  # type: ignore
+    for alloca in callsite_args:
+        if not _pass_via_stack(func_t)[alloca.name]:
+            continue
+        ptr = _alloca_table[alloca._id]
+        stack_arg = bb.append_instruction("mload", ptr)
+        assert stack_arg is not None
+        stack_args.append(stack_arg)
+
+    if returns_count > 0:
+        outs = bb.append_invoke_instruction(stack_args, returns=returns_count)  # type: ignore
+        assert isinstance(return_buf, IROperand)
+        for i, outv in enumerate(outs):
+            if i == 0:
+                dst = return_buf
+            else:
+                ofst = bb.append_instruction("assign", IRLiteral(32 * i))
+                assert ofst is not None
+                new_dst = bb.append_instruction("add", return_buf, ofst)
+                assert new_dst is not None
```

Comment:
why do we need to keep both `dst` and `new_dst`?

Reply:
tried to avoid another type ignore